### ISSUE TITLE
Fatal error on REL1_28

### ISSUE THREAD

**Issue by freephile** [enterprisemediawiki/Wiretap#17]
On https://demo.qualitybox.us/wiki/Special:Wiretap?requestDebug=true&useskin=vector

I get **Fatal error**: Call to undefined method SkinVector::link() in **/opt/htdocs/mediawiki/extensions/Wiretap/SpecialWiretap.php** on line **528**

That line is
```php
$page = $this->getSkin()->link( $pageTitle );
```
I tried inserting a <code>context</code> (property) or <code>getContext()</code> method into the chain, but that too errored:
```php
$page = $this->getContext->getSkin()->link( $pageTitle );
```
I could only find a <code>link</code> method in the Linker class. https://doc.wikimedia.org/mediawiki-core/master/php/classLinker.html#ad97e40cfea2343c5cae8307cf1fe47a4
## Solution
A little more digging reveals that indeed there was a change in REL1_28 that removed this link functionality. All the previously aliased methods in Skin have been removed. You now [have to use LinkRenderer instead](https://gerrit.wikimedia.org/g/mediawiki/core/+/master/HISTORY#938)

Here's an example of a Special Page that was updated for this change.
https://gerrit.wikimedia.org/r/#/c/292077/3/includes/specials/SpecialNewpages.php

**Comment by freephile** [comment]
See https://www.mediawiki.org/wiki/Manual:LinkRenderer
The [documentation for Linker](https://www.mediawiki.org/wiki/Manual:Linker.php) still shows a 'link' method, but the release notes say that you should use LinkRenderer for versions of MediaWiki >= 1.28

There is an old Google Code In project [to clean up extensions](https://codein.withgoogle.com/archive/2016/organization/5709875686408192/task/6470810881490944/) which use the deprecated Linker class methods. This led to a [Phab Task #T149346 with the same goal](https://phabricator.wikimedia.org/T149346).

The [removed methods as of REL1_28](https://gerrit.wikimedia.org/g/mediawiki/core/+/master/HISTORY#938)

Aliases for Linker methods, deprecated since 1.21, were removed from Skin:
* Skin::commentBlock() (use Linker::commentBlock() instead)
* Skin::generateRollback() (use Linker::generateRollback() instead)
* Skin::link() (use MediaWiki\Linker\LinkRenderer instead)
* Skin::linkKnown() (use MediaWiki\Linker\LinkRenderer instead)
* Skin::userLink() (use Linker::userLink() instead)
* Skin::userToolLinks() (use Linker::userToolLinks() instead)

**Comment by jamesmontalvo3** [comment]
Closed by #19
